src/public_key/mod.rs

```diff
@@ -0,0 +1,673 @@
+    fn eq(&self, other: &Self) -> bool {
+        self.master_private_key == other.master_private_key
+            && self.master_public_key == other.master_public_key
+            && self.master_public_delegation_key == other.master_public_delegation_key
+    }
+}
+
+impl<T: AbeScheme + Clone + PartialEq> AsBytes for AbeMasterKey<T> {
+    fn as_bytes(&self) -> Result<Vec<u8>, FormatErr> {
+        let mut bytes = self.master_private_key.as_bytes()?;
+        bytes.extend(
+            self.master_public_key
+                .as_bytes()
+                .map_err(|_e| FormatErr::ConversionFailed)?,
+        );
+        bytes.extend(
+            self.master_public_delegation_key
+                .as_bytes()
+                .map_err(|_e| FormatErr::ConversionFailed)?,
+        );
+        Ok(bytes)
+    }
+
+    fn len_bytes(&self) -> usize {
+        let bytes = self.as_bytes();
+        match bytes {
+            Ok(bytes) => bytes.len(),
+            Err(_e) => 0,
+        }
+    }
+
+    fn from_bytes(bytes: &[u8]) -> Result<Self, FormatErr> {
+        let master_private_key = T::MasterPrivateKey::from_bytes(bytes)
+            .map_err(|e| FormatErr::Deserialization(e.to_string()))?;
+        let master_public_key =
+            T::MasterPublicKey::from_bytes(&bytes[master_private_key.len_bytes()..])
+                .map_err(|e| FormatErr::Deserialization(e.to_string()))?;
+        let master_public_delegation_key = T::MasterPublicDelegationKey::from_bytes(
+            &bytes[master_private_key.len_bytes() + master_public_key.len_bytes()..],
+        )
+        .map_err(|e| FormatErr::Deserialization(e.to_string()))?;
+        Ok(AbeMasterKey {
+            master_private_key,
+            master_public_key,
+            master_public_delegation_key,
+        })
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct AbeCrypto<T: AbeScheme + Clone + PartialEq> {
+    scheme: T,
+    attrs: Option<Vec<u32>>,
+}
+
+impl<T: AbeScheme + Clone + PartialEq> AbeCrypto<T>
+where
+    AbeKeyPair<T>: PartialEq + Clone,
+    AbePrivateKey<T>: PartialEq + Clone,
+    AbePublicKey<T>: PartialEq + Clone,
+{
+    pub fn generate_key_pair(
+        policy: &str,
+        master_priv_key: &<T as AbeScheme>::MasterPrivateKey,
+        master_pub_key: &<T as AbeScheme>::MasterPublicKey,
+    ) -> anyhow::Result<AbeKeyPair<T>> {
+        // warning: in this case, the encryption key is the 'public key'
+        // and the decryption key is the 'private key'
+        let user_decryption_key = T::default()
+            .key_generation(&MonotoneSpanProgram::<i32>::parse(policy)?, master_priv_key)?;
+        Ok(AbeKeyPair {
+            public_key: AbePublicKey(master_pub_key.clone()),
+            private_key: user_decryption_key,
+        })
+    }
+
+    pub fn decrypt_symmetric_key_with_decryption_key<S: SymmetricCrypto>(
+        &self,
+        decryption_key: &AbePrivateKey<T>,
+        data: &[u8],
+    ) -> anyhow::Result<S::Key> {
```

Comment:
Shouldn't be a real result instead of anyhow::Result ?

Reply:
Will be fixed later in https://github.com/Cosmian/abe_gpsw/issues/9